downstairs/src/lib.rs

```diff
@@ -3228,4 +3431,318 @@ mod test {
+            block_size.trailing_zeros(),
+        ));
+        region_options.set_uuid(Uuid::new_v4());
+
+        let dir = tempdir()?;
+        mkdir_for_file(dir.path())?;
+
+        let mut region = Region::create(&dir, region_options)?;
+        region.extend(2)?;
+
+        let path_dir = dir.as_ref().to_path_buf();
+
+        build_downstairs_for_region(
+            &path_dir, false, // lossy
+            false, // return_errors
+            read_only,
+        )
+    }
+
+    #[tokio::test]
+    async fn test_promote_to_active_one_read_write() -> Result<()> {
+        let ads = build_test_downstairs(false)?;
+
+        let upstairs_connection = UpstairsConnection {
+            upstairs_id: Uuid::new_v4(),
+            session_id: Uuid::new_v4(),
+            gen: 1,
+        };
+
+        let (_tx, mut _rx) = channel(1);
+        let tx = Arc::new(_tx);
+
+        let mut ds = ads.lock().await;
+        ds.promote_to_active(upstairs_connection, tx).await?;
+
+        assert_eq!(ds.active_upstairs().len(), 1);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_promote_to_active_one_read_only() -> Result<()> {
+        let ads = build_test_downstairs(true)?;
+
+        let upstairs_connection = UpstairsConnection {
+            upstairs_id: Uuid::new_v4(),
+            session_id: Uuid::new_v4(),
+            gen: 1,
+        };
+
+        let (_tx, mut _rx) = channel(1);
+        let tx = Arc::new(_tx);
+
+        let mut ds = ads.lock().await;
+        ds.promote_to_active(upstairs_connection, tx).await?;
+
+        assert_eq!(ds.active_upstairs().len(), 1);
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_promote_to_active_multi_read_write_different_uuid(
+    ) -> Result<()> {
+        // Attempting to activate multiple read-write (where it's different
+        // Upstairs) should kick out the other connection
+        let ads = build_test_downstairs(false)?;
+
+        let upstairs_connection_1 = UpstairsConnection {
+            upstairs_id: Uuid::new_v4(),
+            session_id: Uuid::new_v4(),
+            gen: 1,
+        };
+
+        let upstairs_connection_2 = UpstairsConnection {
+            upstairs_id: Uuid::new_v4(),
+            session_id: Uuid::new_v4(),
+            gen: 1,
+        };
+
+        let (tx1, mut rx1) = channel(1);
+        let tx1 = Arc::new(tx1);
+
+        let (tx2, mut rx2) = channel(1);
+        let tx2 = Arc::new(tx2);
+
+        let mut ds = ads.lock().await;
+        ds.promote_to_active(upstairs_connection_1, tx1).await?;
+
+        assert_eq!(ds.active_upstairs().len(), 1);
+        assert!(matches!(rx1.try_recv().err().unwrap(), TryRecvError::Empty));
+        assert!(matches!(rx2.try_recv().err().unwrap(), TryRecvError::Empty));
+
+        ds.promote_to_active(upstairs_connection_2, tx2).await?;
+        assert_eq!(rx1.try_recv().unwrap(), upstairs_connection_2);
+        assert!(matches!(rx2.try_recv().unwrap_err(), TryRecvError::Empty));
+
```

Comment:
Could we also test `is_active` in here?  Making sure `upstairs_connection_1` is not active and `upstairs_connection_2` is?

Reply:
Good idea, done in 39514d5